daffodil-runtime1/src/main/scala/org/apache/daffodil/dpath/ConverterOps.scala

```diff
@@ -23,74 +23,158 @@ import org.apache.daffodil.calendar.DFDLDate
+import org.apache.daffodil.infoset.DataValue.DataValueShort
+import org.apache.daffodil.infoset.DataValue.DataValueString
+import org.apache.daffodil.infoset.DataValue.DataValueBigDecimal
+import org.apache.daffodil.infoset.DataValue.DataValueString
+import org.apache.daffodil.infoset.DataValue.DataValueDouble
+import org.apache.daffodil.infoset.DataValue.DataValueBigInt
+import org.apache.daffodil.infoset.DataValue.DataValueAny
+import org.apache.daffodil.infoset.DataValue.DataValueBool
+import org.apache.daffodil.infoset.DataValue.DataValuePrimitive
+import org.apache.daffodil.infoset.DataValue.DataValuePrimitive
+import org.apache.daffodil.infoset.DataValue.DataValuePrimitive
+import org.apache.daffodil.infoset.DataValue.DataValueDate
+import org.apache.daffodil.infoset.DataValue.DataValueDouble
+import org.apache.daffodil.infoset.DataValue.DataValueDouble
 
 case object BooleanToLong extends Converter {
-  override def computeValue(a: AnyRef, dstate: DState): AnyRef = JLong.valueOf(if (asBoolean(a) == true) 1L else 0L)
+  override def computeValue(a: DataValuePrimitive, dstate: DState): DataValueLong = JLong.valueOf(if (asBoolean(a) == true) 1L else 0L)
 }
 
 case object BooleanToString extends Converter {
-  override def computeValue(a: AnyRef, dstate: DState): AnyRef = if (asBoolean(a) == true) "true" else "false"
+  override def computeValue(a: DataValuePrimitive, dstate: DState): DataValueString = if (asBoolean(a) == true) "true" else "false"
 }
 
 case object DateTimeToDate extends Converter {
-  override def computeValue(a: AnyRef, dstate: DState): AnyRef = {
-    a match {
+  override def computeValue(a: DataValuePrimitive, dstate: DState): DataValueDate = {
+    a.getAnyRef match {
       case dt: DFDLDateTime => dt.toDate
       case _ => throw new NumberFormatException("xs:dateTime expected but an invalid type was received.")
     }
   }
 }
 case object DateTimeToTime extends Converter {
-  override def computeValue(a: AnyRef, dstate: DState): AnyRef = {
-    a match {
+  override def computeValue(a: DataValuePrimitive, dstate: DState): DataValueTime = {
+    a.getAnyRef match {
       case dt: DFDLDateTime => dt.toTime
       case _ => throw new NumberFormatException("xs:dateTime expected but an invalid type was received.")
     }
   }
 }
 case object DateToDateTime extends Converter {
-  override def computeValue(a: AnyRef, dstate: DState): AnyRef = {
-    a match {
+  override def computeValue(a: DataValuePrimitive, dstate: DState): DataValueDateTime = {
+    a.getAnyRef match {
       case d: DFDLDate => d.toDateTime
       case _ => throw new NumberFormatException("xs:date expected but an invalid type was received.")
     }
   }
 }
 case object DecimalToInteger extends Converter {
-  override def computeValue(a: AnyRef, dstate: DState): AnyRef = asBigDecimal(a).toBigInteger()
+  override def computeValue(a: DataValuePrimitive, dstate: DState): DataValueBigInt = asBigDecimal(a.getAnyRef).toBigInteger()
```

Comment:
All of this asType(x.getAnyRef) stuff could be cleaned up as well.

Because of how our libraries are structured, the asType functions do not have access to the DataValue type.

We could probably move the functions (or wrappers for them) to the DataValue class, so we have asFoo() methods which do intellegent coercien to complement our getFoo() methods which just assume the underlying value is already the correct type.

I'm not convinced this is worth the effort at this point.